### PR TITLE
Add thumbnails for files in the activity view

### PR DESCRIPTION
**Pull request by claucambra**
This PR adds thumbnails for file changes in the activity list.

image

Partially addresses #3666

### CODE REVIEW

File: src/gui/tray/ActivityItem.qml
Comment (CarlSchwan) [inline src/gui/tray/ActivityItem.qml:55]: Can you make sure that the cache works across restart? Generating a thumbnail is a bit expensive on the server side.

**Review by CarlSchwan — COMMENTED** [review]
Cool feature. Could you also test it with previews disabled from the server side?

**Comment by claucambra** [comment]
> Cool feature. Could you also test it with previews disabled from the server side?

Yeah, instead of previews the preview API presents thumbnails corresponding to the file type:

Screenshot from 2022-01-24 15-21-41

File: src/gui/tray/ActivityItem.qml
Comment (claucambra) [reply src/gui/tray/ActivityItem.qml:55]: Eyeballing it, it seems to work across computer restarts but I'm not sure, is there some way to tell if QML is loading the image from cache/if the server is receiving a GET request for the thumbnail>

**Comment by jancborchardt** [comment]
Yeah, really great! Only a small enhancement: Could you give the image thumbnails 3px border radius so it's the same as on server?

File: src/gui/tray/ActivityItem.qml
Comment (mgallien) [reply src/gui/tray/ActivityItem.qml:55]: @CarlSchwan are those previews cached on server side ?

**Comment by mgallien** [comment]
@claucambra this PR needs a rebase to solve the existing conflicts

**Comment by claucambra** [comment]
> Yeah, really great! Only a small enhancement: Could you give the image thumbnails 3px border radius so it's the same as on server?

Sure, here's a screenshot of the change:

image

File: src/gui/systray.cpp
Comment (mgallien) [inline src/gui/systray.cpp:69]: why twice the same object used to create two different image providers ?
what is the benefit of duplication ?

File: src/gui/tray/ActivityItem.qml
Comment (mgallien) [reply src/gui/tray/ActivityItem.qml:56]: so the cache is not kept after a restart
I always see that we download the preview images

File: src/gui/systray.cpp
Comment (claucambra) [reply src/gui/systray.cpp:69]: Missed this, going to unify them

Comment (claucambra) [reply src/gui/systray.cpp]: Done!

File: src/gui/tray/activitydata.cpp
Comment (mgallien) [inline src/gui/tray/activitydata.cpp:31]: we do not add indentation inside a namespace
you have extra spaces

File: src/gui/tray/activitylistmodel.cpp
Comment (mgallien) [inline src/gui/tray/activitylistmodel.cpp:219]: why not `a_previews.empty()` ?

File: src/gui/tray/activitylistmodel.cpp
Comment (mgallien) [inline src/gui/tray/activitylistmodel.cpp:283]: same question

File: src/gui/tray/activitylistmodel.cpp
Comment (mgallien) [inline src/gui/tray/activitylistmodel.cpp:338]: `const QJsonArray &activities`
why not passing by const reference ?

File: src/gui/tray/usermodel.cpp
Comment (mgallien) [inline src/gui/tray/usermodel.cpp:570]: extra empty line

File: src/gui/tray/usermodel.cpp
Comment (mgallien) [inline src/gui/tray/usermodel.cpp:511]: please use const references
(and I know QString are ref counted but still)

File: test/testactivitydata.cpp
Comment (mgallien) [inline test/testactivitydata.cpp:134]: can you make this test a data driven one ?
please have a look at this tutorial from [Qt doc](https://doc.qt.io/qt-5/qttestlib-tutorial2-example.html)

**Review by mgallien — CHANGES_REQUESTED** [review]

File: src/gui/tray/ActivityItem.qml
Comment (mgallien) [inline src/gui/tray/ActivityItem.qml:63]: is this `Item` really needed ?
cannot you directly instantiate the `Rectangle` ?

**Review by camilasan — COMMENTED** [review]
> Merge branch 'master' into feature/activity_thumbnails 
Please rebase, not merge.

**Review by jancborchardt — CHANGES_REQUESTED** [review]
Good stuff @claucambra, just 2 more things:
- The thumbnails seem a bit small, could you increase the size of them so that their left edge lines up with the left edge of the search field (of course they should still be centered, so they will also extend to the right as well as top and bottom)
   - Maybe to make it look nice, the mimetype icons also have to be made larger a bit if possible – or is that their native pixel size from the SVG so they don’t get blurry?
   - (In the screenshot below, the "Tasks" icon for example looks good, and the thumbnails too small)
- All text should left-align with the username in the header, so the "All synced" status and Search field text needs to move a bit right, and the text of the entries a bit to the left

desktop tray thumbnails

File: src/gui/tray/activitylistmodel.cpp
Comment (allexzander) [inline src/gui/tray/activitylistmodel.cpp:169]: Why not `const PreviewData &`?

File: src/gui/tray/activitylistmodel.cpp
Comment (allexzander) [inline src/gui/tray/activitylistmodel.cpp:345]: `for (const auto &activ`

File: src/gui/tray/activitylistmodel.cpp
Comment (allexzander) [inline src/gui/tray/activitylistmodel.cpp:346]: Would be nice to add `const` here

Comment (allexzander) [inline src/gui/tray/activitylistmodel.cpp:346]: Would be nice to add const here

File: src/gui/tray/activitylistmodel.cpp
Comment (allexzander) [inline src/gui/tray/activitylistmodel.cpp:367]: `QStringLiteral("ocs")` and `QStringLiteral("data")`

File: src/gui/tray/activitylistmodel.cpp
Comment (allexzander) [inline src/gui/tray/activitylistmodel.cpp:369]: Why not just `if (_accountState)`? Also, why not `const auto ast`? Is this check even needed here with no immediate usage of `_accountState`?

File: src/gui/tray/usermodel.cpp
Comment (allexzander) [inline src/gui/tray/usermodel.cpp:6]: I wouldn't change that, `pushnotifications.h` is in `libsync`

File: src/gui/tray/usermodel.cpp
Comment (allexzander) [inline src/gui/tray/usermodel.cpp:512]: Why remove `QStringLiteral()` around literals?

File: src/gui/tray/usermodel.cpp
Comment (allexzander) [inline src/gui/tray/usermodel.cpp:573]: ?

File: src/gui/tray/usermodel.h
Comment (allexzander) [inline src/gui/tray/usermodel.h:128]: Please add underscore prefix as in other private members

File: src/gui/tray/activitydata.cpp
Comment (claucambra) [reply src/gui/tray/activitydata.cpp]: Got it, removed

File: src/gui/tray/activitylistmodel.cpp
Comment (claucambra) [reply src/gui/tray/activitylistmodel.cpp]: No good reason! Replaced with `empty()`

File: src/gui/tray/activitylistmodel.cpp
Comment (claucambra) [reply src/gui/tray/activitylistmodel.cpp]: Replaced

File: src/gui/tray/activitylistmodel.cpp
Comment (claucambra) [reply src/gui/tray/activitylistmodel.cpp]: Now passed by reference

File: src/gui/tray/usermodel.cpp
Comment (claucambra) [reply src/gui/tray/usermodel.cpp]: Removed

File: src/gui/tray/usermodel.cpp
Comment (claucambra) [reply src/gui/tray/usermodel.cpp]: Thanks, QStrings now passed by reference

File: test/testactivitydata.cpp
Comment (claucambra) [reply test/testactivitydata.cpp]: I have used the column and row structure from the documentation now, thanks

File: src/gui/tray/ActivityItem.qml
Comment (claucambra) [reply src/gui/tray/ActivityItem.qml]: Simplified with rectangle now

File: src/gui/tray/activitylistmodel.cpp
Comment (mgallien) [reply src/gui/tray/activitylistmodel.cpp:404]: the `ast` name is probably a bad abbreviation
seeing the comment from @allexzander I was pretty sure we were doing some kind of abstract syntax trees
as a general rule we should avoid abbreviations as they are not easy to read later (or by somebody else)

File: src/gui/tray/activitylistmodel.cpp
Comment (claucambra) [reply src/gui/tray/activitylistmodel.cpp]: Thanks, fixed now

File: src/gui/tray/activitylistmodel.cpp
Comment (claucambra) [reply src/gui/tray/activitylistmodel.cpp]: True, added

File: src/gui/tray/activitylistmodel.cpp
Comment (claucambra) [reply src/gui/tray/activitylistmodel.cpp]: Now use QStringLiterals

File: src/gui/tray/activitylistmodel.cpp
Comment (claucambra) [reply src/gui/tray/activitylistmodel.cpp]: This was part of the code before and I hadn't changed it, but yes it is redundant so I have removed it now, thanks

File: src/gui/tray/usermodel.cpp
Comment (claucambra) [reply src/gui/tray/usermodel.cpp]: Sorry, reverted

File: src/gui/tray/usermodel.cpp
Comment (claucambra) [reply src/gui/tray/usermodel.cpp]: I can add QStringLiterals around the conditionals but the QObject::tr only accepts char pointers so those have to stay as literals

File: src/gui/tray/usermodel.cpp
Comment (claucambra) [reply src/gui/tray/usermodel.cpp:572]: Removed this empty line

File: src/gui/tray/usermodel.h
Comment (claucambra) [reply src/gui/tray/usermodel.h]: Added, thanks

**Comment by claucambra** [comment]
> Good stuff @claucambra, just 2 more things:
> 
>     * The thumbnails seem a bit small, could you increase the size of them so that their left edge lines up with the left edge of the search field (of course they should still be centered, so they will also extend to the right as well as top and bottom)
>       
>       * Maybe to make it look nice, the mimetype icons also have to be made larger a bit if possible – or is that their native pixel size from the SVG so they don’t get blurry?
>       * (In the screenshot below, the "Tasks" icon for example looks good, and the thumbnails too small)
> 
>     * All text should left-align with the username in the header, so the "All synced" status and Search field text needs to move a bit right, and the text of the entries a bit to the left
> 
> 
> desktop tray thumbnails

Here's a screenshot with the new sizes, also generally aligned things in the tray:

image

**Comment by allexzander** [comment]
Nice. I'd approve after Jan gives a go design-wise

**Comment by claucambra** [comment]
> Nice. I'd approve after Jan gives a go design-wise

@jancborchardt is there anything else I should address in this PR? :)

**Comment by mgallien** [comment]
@claucambra can you rebase it on top of master ?
there are conflicts

File: src/gui/tray/ActivityItemContent.qml
Comment (mgallien) [inline src/gui/tray/ActivityItemContent.qml:37]: `model.thumbnail` can be returned undefined in that case every usage of it in this component will generate a warning
can you check if you can use the same logic that is used in `activityIcon` ?

File: src/gui/tray/activitylistmodel.h
Comment (mgallien) [inline src/gui/tray/activitylistmodel.h:68]: this is missing a trailing comma

File: src/gui/tray/activitylistmodel.cpp
Comment (mgallien) [inline src/gui/tray/activitylistmodel.cpp:304]: see my comment in qml linked to this empty return

**Review by mgallien — COMMENTED** [review]
a few comments and we should be good to go
sorry for the delay

File: src/gui/tray/activitylistmodel.cpp
Comment (claucambra) [reply src/gui/tray/activitylistmodel.cpp:304]: Got it, fixed in QML

File: src/gui/tray/ActivityItemContent.qml
Comment (claucambra) [reply src/gui/tray/ActivityItemContent.qml]: Good catch, I added a check for `model.thumbnail` not being undefined to set the property

File: src/gui/tray/activitylistmodel.h
Comment (claucambra) [reply src/gui/tray/activitylistmodel.h]: Adde

File: src/gui/tray/ActivityItemContent.qml
Comment (mgallien) [inline src/gui/tray/ActivityItemContent.qml:37]: I still get the warnings
`
2022-03-09 12:49:43:798 [ warning default qrc:/qml/src/gui/tray/ActivityItemContent.qml:37 ]:   qrc:/qml/src/gui/tray/ActivityItemContent.qml:37:13: Unable to assign [undefined] to QUrl
2022-03-09 12:49:43:798 [ warning default qrc:/qml/src/gui/tray/ActivityItemContent.qml:33 ]:   qrc:/qml/src/gui/tray/ActivityItemContent.qml:33: TypeError: Value is undefined and could not be converted to an object
2022-03-09 12:49:43:798 [ warning default qrc:/qml/src/gui/tray/ActivityItemContent.qml:32 ]:   qrc:/qml/src/gui/tray/ActivityItemContent.qml:32: TypeError: Value is undefined and could not be converted to an object
 `
this is partially due to me having my dev nextcloud certificate being rejected by Image qml network code
still I think if we are going to access properties inside `model.thumbnail` maybe you could always create them when you return `thumbnail` role in the model

File: src/gui/tray/activitydata.cpp
Comment (mgallien) [inline src/gui/tray/activitydata.cpp:88]: can you improve the name of variable `i` ?

**Review by mgallien — COMMENTED** [review]
small extra comments

File: src/gui/tray/activitylistmodel.cpp
Comment (claucambra) [reply src/gui/tray/activitylistmodel.cpp]: Thanks, changed

File: src/gui/tray/activitydata.cpp
Comment (claucambra) [reply src/gui/tray/activitydata.cpp]: Yes, now `subjectRichParameterMatch`

**Comment by mgallien** [comment]
@claucambra can you clean history before approval ?

**Comment by claucambra** [comment]
> @claucambra can you clean history before approval ?

Yup, squashed

**Review by mgallien — APPROVED** [review]

**Review by allexzander — APPROVED** [review]

**Review by jancborchardt — APPROVED** [review]
Looks all good! :)

**Pull request merged**
